src/main/java/jnr/ffi/Platform.java

```diff
@@ -500,32 +511,81 @@ public String mapLibraryName(String libName) {
             }
             return "lib" + libName + ".dylib";
         }
-        
-        @Override
-        public String getName() {
-            return "Darwin";
-        }
 
     }
+
     /**
      * A {@link Platform} subclass representing the Linux operating system.
      */
     static final class Linux extends Supported {
 
+        // represents a valid library file that matches the search
+        private static class Match implements Comparable<Match> {
+            String path; // absolute path of library file
+            int[] version; // version of library, empty if no version specified
+            boolean isCustom; // if path is from a custom searchPath specified by user and not default path
+
+            @Override
+            public int compareTo(Match o) { // for Collections.sort() to work
+                return compareVersions(o.version, this.version);
+            }
+        }
+
         public Linux() {
             super(OS.LINUX);
         }
 
         @Override
-        public String locateLibrary(final String libName, List<String> libraryPaths) {
+        public String locateLibrary(String libName, List<String> libraryPaths) {
+            return locateLibrary(libName, libraryPaths, null);
+        }
+
+        @Override
+        public String locateLibrary(final String libName, List<String> libraryPaths,
+                                    Map<LibraryOption, Object> options) {
+            List<Match> matches = getMatches(libName, libraryPaths);
+            if (matches.isEmpty()) return mapLibraryName(libName); // no matches, default behavior returns mapped name
+
+            boolean preferCustom = options != null && options.containsKey(LibraryOption.PreferCustomPaths);
+
+            Collections.sort(matches); // sort by version, regardless of location
+
+            Match best = null;
+            if (preferCustom) {
+                for (Match match : matches) {
+                    if (match.isCustom) {
+                        best = match;
+                        break;
+                    }
+                }
+            }
+            return best != null ? best.path : matches.get(0).path;
+        }
+
+        private List<Match> getMatches(String libName, List<String> libraryPaths) {
+            List<String> customPaths = new ArrayList<>();
+            if (LibraryLoader.DefaultLibPaths.PATHS.size() > 0 &&
+                    libraryPaths.size() >= LibraryLoader.DefaultLibPaths.PATHS.size()) {
```

Comment:
Not sure I get why the libraryPaths size is checked against the default size

Reply:
This is for the following edge case scenario that I want us to support:

Caller explicitly uses one of the default system paths as a custom path, for example they explicitly want `/lib/x86_64-linux-gnu/` instead of `/usr/lib/`. 

In this case the caller *prefers* a system path, our logic would fail for this scenario without this additional code, and would end up picking whichever one came first even though the caller **specifically requested** a certain path that *happened* to be a system one

I agree it's a little over-engineering, but it makes the behavior much more strictly defined, that is:

If you use `LibraryOption.PreferCustomPaths` your custom paths will ***always*** be preferred, no matter what or where they actually are. Otherwise, we will pick for you the highest version we find first.